platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,493 @@
+- Efficient Storage: Roster data should be stored efficiently in the state to minimize overhead and unnecessary data
+  copying.
+- Clear API: A well-defined API should be provided for the App to create and submit candidate rosters to the platform,
+  and for the
+  platform to manage the lifecycle of submitted rosters.
+- Components update: Platform components that currently rely on the use of the Address book (such as Reconnect, Network,
+  Event Validation etc.) must be adapted to use the new `Roster` object instead.
+- Address book paradigm discontinued within the platform codebase in favor of `Roster`s
+
+## Architecture
+
+###### Roster Lifecycle
+
+![](TSS%20Roster%20Lifecycle.drawio.svg)
+
+### Roster Public API
+
+#### Setting the Candidate Roster
+
+A new method will be added to the platform state to allow the App to submit a Candidate Roster to the platform:
+
+```java
+//in PlaformState
+
+void setCandidateRoster(@NonNull final Roster candidateRoster);
+```
+
+The steps to validate and add the Roster to the state will be executed within the same thread that submits the request.
+An exception will be thrown if the Roster is invalid.
+The durability of the submitted Roster is ensured via storing it in the state.
+
+The existing `PlatformState` class is in the middle of a refactor, and in the long term, may cease to exist.
+However, its replacement will continue to have this new API to set a candidate roster.
+
+## Data Structure
+
+The App is already responsible for managing the address book.
+We propose that it continues to do so, and when it receives a HAPI transaction, creates a candidate Roster object
+and passes it to the platform via the new API. The platform will then validate the roster and store it in the state.
+
+### Protobuf
+
+#### Roster
+
+The protobuf definition for the Roster will look as follows:
+
+```proto
+/**
+ * A single roster in the network state.
+ * <p>
+ * The roster SHALL be a map of key `NodeId`, value `RosterEntry`.
+ */
+message Roster {
+
+    /**
+     * Map of roster entries, one per consensus node.
+     * <p>
+     * This map SHALL contain roster entries in natural order of ascending node ids.
+     * The node Id key SHALL match the node Id in the value.
+     * This map SHALL NOT be empty.<br/>
+     */
+    map<uint64, RosterEntry> rosterEntries = 1;
+}
+```
+
+#### RosterEntry
+
+A roster entry will look as follows:
+
+```proto
+/**
+ * A single roster entry in the network state.
+ *
+ * Each roster entry in SHALL encapsulate the elements required
+ * to manage node participation in the Threshold Signature Scheme (TSS).<br/>
+ * All fields are REQUIRED.
+ */
+message RosterEntry {
+
+    /**
+     * A consensus node identifier.
+     * <p>
+     * Node identifiers SHALL be unique _within_ a shard and realm,
+     * but a node SHALL NOT, ever, serve multiple shards or realms,
+     * therefore the node identifier MAY be repeated _between_ shards and realms.
+     */
+    uint64 node_id = 1;
```

Comment:
Node ID's are intended to be _ledger global_, not scoped within a shard and/or realm.

Reply:
This doc was lifted from the existing protobuf definition of `node_id` so I suspect we have this error everywhere. @jsync-swirlds, you might find this of note.

Reply:
Noted.
We'll modify this the next time we update the node protobuf files. Given we don't have shards or realms and haven't figured out how the counter might be communicated across shards and realms, it doesn't seem worth a separate HIP just for this item, though.

Reply:
Let's please update the language here so it is correct, and then we can fix this in the other protobuf files separately.

Reply:
updated.